PayChannel ledger node - 1st draft

mDuo13 opened the pull request:
Companion to #229. This one covers the new ledger node type added by the PayChan amendment. The RPC reference, tutorial, and concept pages are still in the pipeline. Cross-references are to be added by another PR after all those are merged.

mDuo13 committed:
paychannel ledger node - 1st draft

seelabs commented on content/reference-ledger-format.md:
```diff
@@ -303,6 +305,72 @@ The `index` of an Offer node is the SHA-512Half of the following values put toge
+
+The `PayChannel` node type represents a payment channel, which is a structure that makes it easier to perform asynchronous, rapid payments and micropayments of XRP. A payment channel holds a balance of XRP that can only be paid out to a specific destination address until the channel is closed. Any unspent XRP is returned to the channel's owner (the source address that created and funded it) when  ...
+
+The `PaymentChannelCreate` transaction type creates a `PayChannel` node. The PaymentChannelFund and PaymentChannelClaim transaction types modify existing `PayChannel` nodes.
```
Nit: If we put `PaymentChannelCreate` in quotes shouldn't we put the `Fund` and `Claim` txs in quotes as well?

seelabs commented on content/reference-ledger-format.md:
```diff
@@ -303,6 +305,72 @@ The `index` of an Offer node is the SHA-512Half of the following values put toge
+
+_(Requires the [PayChan Amendment](concept-amendments.html#paychan).)_
+
+The `PayChannel` node type represents a payment channel, which is a structure that makes it easier to perform asynchronous, rapid payments and micropayments of XRP. A payment channel holds a balance of XRP that can only be paid out to a specific destination address until the channel is closed. Any unspent XRP is returned to the channel's owner (the source address that created and funded it) when  ...
```
The description doesn't mention anything about the off-ledger nature of channels. How about:
Payment channels permit off-ledger checkpoints of XRP payments flowing in a single direction. (that's from the code).

mDuo13 replied:
I think that description from the code is still a little obtuse. Here's my attempt at a clearer wording:

> The `PayChannel` node type represents a payment channel. Payment channels enable small, rapid off-ledger payments of XRP that can be later reconciled with the consensus ledger.

seelabs commented on content/reference-ledger-format.md:
```diff
@@ -303,6 +305,72 @@ The `index` of an Offer node is the SHA-512Half of the following values put toge
+
+The `PaymentChannelCreate` transaction type creates a `PayChannel` node. The PaymentChannelFund and PaymentChannelClaim transaction types modify existing `PayChannel` nodes.
+
+When a payment channel expires, at first it remains on the ledger, because only new transactions can modify ledger contents. Transaction processing automatically closes a payment channel when any transaction that accesses it after the expiration. Therefore, to "finalize" the closing of an expired channel and return the unspent XRP to the owner, some address must send a new PaymentChannelClaim or  ...
```
"any transaction that accesses" -> "any transaction accesses"

seelabs commented on content/reference-ledger-format.md:
```diff
@@ -303,6 +305,72 @@ The `index` of an Offer node is the SHA-512Half of the following values put toge
+| `Account`           | String    | AccountID         | The source address that owns this payment channel. This comes from the sending address of the transaction that created the channel. |
+| `Destination`       | String    | AccountID         | The destination address for this payment channel. While the payment channel is open, this address is the only one that can receive XRP from the channel. This comes from the `Destination` field of the transaction that created the channel. |
+| `Amount`            | String    | Amount            | Total XRP, in drops, that has been allocated to this channel. This includes XRP that has been paid to the destination address. This is initially set by the transaction that created the channel and can be increased if the source address sends a PaymentChannelFund transaction. |
+| `Balance`           | String    | Amount            | Total XRP, in drops, currently stored in the channel. This XRP can still be paid to the destination address with PaymentChannelClaim transactions. If the channel closes, this XRP is returned to the source address. |
```
That's not right. `Balance` is the total amount paid to the destination so far. If the channel closes then `Amount - Balance` is returned to the source address.

seelabs commented on content/reference-ledger-format.md:
```diff
@@ -303,6 +305,72 @@ The `index` of an Offer node is the SHA-512Half of the following values put toge
+| `Destination`       | String    | AccountID         | The destination address for this payment channel. While the payment channel is open, this address is the only one that can receive XRP from the channel. This comes from the `Destination` field of the transaction that created the channel. |
+| `Amount`            | String    | Amount            | Total XRP, in drops, that has been allocated to this channel. This includes XRP that has been paid to the destination address. This is initially set by the transaction that created the channel and can be increased if the source address sends a PaymentChannelFund transaction. |
+| `Balance`           | String    | Amount            | Total XRP, in drops, currently stored in the channel. This XRP can still be paid to the destination address with PaymentChannelClaim transactions. If the channel closes, this XRP is returned to the source address. |
+| `PublicKey`         | String    | PubKey            | Public key, in hexadecimal, of the key pair that can be used to sign claims against this channel. This can be any valid secp256k1 or Ed25519 public key. This is set by the transaction that created the channel and must match the public key used in claims against the channel. The channel source address can also send XRP to the destination with ...
```
Nit: `send XRP` -> `send XRP from the channel`

seelabs commented on content/reference-ledger-format.md:
```diff
@@ -303,6 +305,72 @@ The `index` of an Offer node is the SHA-512Half of the following values put toge
+| `Balance`           | String    | Amount            | Total XRP, in drops, currently stored in the channel. This XRP can still be paid to the destination address with PaymentChannelClaim transactions. If the channel closes, this XRP is returned to the source address. |
+| `PublicKey`         | String    | PubKey            | Public key, in hexadecimal, of the key pair that can be used to sign claims against this channel. This can be any valid secp256k1 or Ed25519 public key. This is set by the transaction that created the channel and must match the public key used in claims against the channel. The channel source address can also send XRP to the destination with ...
+| `SettleDelay`       | Number    | UInt32            | Number of seconds the source address must wait to close the channel if it still has any XRP in it. Smaller values mean that the destination address has less time to redeem any outstanding claims after the source address requests to close the channel. Can be any value that fits in a 32-bit unsigned integer (0 to 2^32-1). This is set by the tr ...
+| `OwnerNode`         | String    | UInt64            | A hint indicating which page of the source address's owner directory links to this node, in case the directory consists of multiple nodes. |
```
`multiple nodes` -> `multiple pages`

seelabs commented on content/reference-ledger-format.md:
```diff
@@ -303,6 +305,72 @@ The `index` of an Offer node is the SHA-512Half of the following values put toge
+| `PreviousTxnID`     | String    | Hash256           | The identifying hash of the transaction that most recently modified this node. |
+| `PreviousTxnLgrSeq` | Number    | UInt32            | The [index of the ledger](#ledger-index) that contains the transaction that most recently modified this node. |
+| `Flags`             | Number    | UInt32            | A bit-map of boolean flags enabled for this payment channel. Currently, the protocol defines no flags for `PayChannel` nodes. |
+| `Expiration`        | Number    | UInt32            | _(Optional)_ The mutable expiration time for this payment channel, in [seconds since the Ripple Epoch](reference-rippled.html#specifying-time). The channel is expired if this value is present and smaller than the previous ledger's [`close_time` field](#header-format). This starts out omitted, and can be added, removed, or updated by the sour ...
```
Setting the `Expiration` from the `ChannelFund` tx works as follows:
The requested expiration must be greater than the minimum of the current `Expiration` (if any) and the `prev_close + settle_delay`. Notice this means you can set an expiration less than `prev_clost + settle_delay`, but only if `Expiration` has already been set. If the tx tries to set an invalid expiration, the tx fails with `temBAD_EXPIRATION`.

The rules for setting `Expriation` in `Claim` are as follows:
1) Calculate a proposed expiration as `prev_close + settle_delay`.
2) If the proposed expiration is less than the current expiration (if any), use the proposed expiration, otherwise keep the current expiration.

ryangyoung commented on content/reference-ledger-format.md:
```diff
@@ -303,6 +305,72 @@ The `index` of an Offer node is the SHA-512Half of the following values put toge
+
+| Name                | JSON Type | [Internal Type][] | Description            |
+|:--------------------|:----------|:------------------|:-----------------------|
+| `LedgerEntryType`   | String    | UInt16            | The value `0x78`, mapped to the string `PayChannel`, indicates that this node is an payment channel object. |
```
"an payment channel object" --> "a payment channel object"

ryangyoung commented on content/reference-ledger-format.md:
```diff
@@ -303,6 +305,72 @@ The `index` of an Offer node is the SHA-512Half of the following values put toge
+| `LedgerEntryType`   | String    | UInt16            | The value `0x78`, mapped to the string `PayChannel`, indicates that this node is an payment channel object. |
+| `Account`           | String    | AccountID         | The source address that owns this payment channel. This comes from the sending address of the transaction that created the channel. |
+| `Destination`       | String    | AccountID         | The destination address for this payment channel. While the payment channel is open, this address is the only one that can receive XRP from the channel. This comes from the `Destination` field of the transaction that created the channel. |
+| `Amount`            | String    | Amount            | Total XRP, in drops, that has been allocated to this channel. This includes XRP that has been paid to the destination address. This is initially set by the transaction that created the channel and can be increased if the source address sends a PaymentChannelFund transaction. |
```
Should "PaymentChannelFund" be in backticks? What is and isn't in backticks seems to be a little inconsistent and should probably be standardized.

mDuo13 replied:
The rule I've been following is that ledger node types need to be in backticks but transaction types don't. It's maybe not the best rule, but that's (mostly) the way the docs are currently written. I'm making a couple corrections to this doc to clarify that.

ryangyoung commented on content/reference-ledger-format.md:
```diff
@@ -303,6 +305,72 @@ The `index` of an Offer node is the SHA-512Half of the following values put toge
+| `LedgerEntryType`   | String    | UInt16            | The value `0x78`, mapped to the string `PayChannel`, indicates that this node is an payment channel object. |
+| `Account`           | String    | AccountID         | The source address that owns this payment channel. This comes from the sending address of the transaction that created the channel. |
+| `Destination`       | String    | AccountID         | The destination address for this payment channel. While the payment channel is open, this address is the only one that can receive XRP from the channel. This comes from the `Destination` field of the transaction that created the channel. |
+| `Amount`            | String    | Amount            | Total XRP, in drops, that has been allocated to this channel. This includes XRP that has been paid to the destination address. This is initially set by the transaction that created the channel and can be increased if the source address sends a PaymentChannelFund transaction. |
```
> Total XRP, in drops

Maybe add a link to https://ripple.com/build/rippled-apis/#specifying-currency-amounts?

ryangyoung commented on content/reference-ledger-format.md:
```diff
@@ -303,6 +305,72 @@ The `index` of an Offer node is the SHA-512Half of the following values put toge
+| `Amount`            | String    | Amount            | Total XRP, in drops, that has been allocated to this channel. This includes XRP that has been paid to the destination address. This is initially set by the transaction that created the channel and can be increased if the source address sends a PaymentChannelFund transaction. |
+| `Balance`           | String    | Amount            | Total XRP, in drops, currently stored in the channel. This XRP can still be paid to the destination address with PaymentChannelClaim transactions. If the channel closes, this XRP is returned to the source address. |
+| `PublicKey`         | String    | PubKey            | Public key, in hexadecimal, of the key pair that can be used to sign claims against this channel. This can be any valid secp256k1 or Ed25519 public key. This is set by the transaction that created the channel and must match the public key used in claims against the channel. The channel source address can also send XRP to the destination with ...
+| `SettleDelay`       | Number    | UInt32            | Number of seconds the source address must wait to close the channel if it still has any XRP in it. Smaller values mean that the destination address has less time to redeem any outstanding claims after the source address requests to close the channel. Can be any value that fits in a 32-bit unsigned integer (0 to 2^32-1). This is set by the tr ...
```
> This is set by the transaction that creates the channel.

Perhaps add a link to that transaction.

mDuo13 replied:
This is to be added after both this and #229 are merged.

ryangyoung reviewed (approved):
A few nitpicks, but nothing that prevents merging this, even if they're unaddressed.

mDuo13 committed:
paychan ledger format - edits per reviews; new section clarifying expiration

mDuo13 commented:
I think I've addressed all the feedback, but I'm going to leave this open until one or both reviewers get a chance to look at the new "Setting Channel Expiration" section I added.

ryangyoung commented on content/reference-ledger-format.md:
```diff
@@ -303,6 +305,97 @@ The `index` of an Offer node is the SHA-512Half of the following values put toge
+
+The PaymentChannelCreate transaction type creates a `PayChannel` node. The PaymentChannelFund and PaymentChannelClaim transaction types modify existing `PayChannel` nodes.
+
+When a payment channel expires, at first it remains on the ledger, because only new transactions can modify ledger contents. Transaction processing automatically closes a payment channel when any transaction accesses it after the expiration. Therefore, to "finalize" the closing of an expired channel and return the unspent XRP to the owner, some address must send a new PaymentChannelClaim or Payme ...
```
> When a payment channel expires,

Add a link to the new expiration section.

ryangyoung commented on content/reference-ledger-format.md:
```diff
@@ -303,6 +305,97 @@ The `index` of an Offer node is the SHA-512Half of the following values put toge
+
+### Setting Channel Expiration
+
+The `Expiration` field of a payment channel is the mutable expiration time, in contrast to the immutable expiration time represented by the `CancelAfter` field. The `Expiration` field is always omitted when a `PayChannel` node is created. There are several ways the `Expiration` field of a `PayChannel` node can be updated. Generally, the source address can set the `Expiration` of a channel freely  ...
```
1. Depending on what you intend to include in the planned conceptual section, it might be helpful to include a brief explanation/expansion of how expirations work in relation to payment channels. I'm thinking of something fairly simple, like mentioning that payment channels are temporary and that there are n ways to control their expiration. 

2. Some kind of diagram (perhaps something like a UML state diagram) would probably go a long way to explaining the different options and the lifecycle of a channel.

mDuo13 replied:
That's a good idea. I've noted that in the task for the concept article (DOC-696)

ryangyoung commented on content/reference-ledger-format.md:
```diff
@@ -303,6 +305,97 @@ The `index` of an Offer node is the SHA-512Half of the following values put toge
+
+### Setting Channel Expiration
+
+The `Expiration` field of a payment channel is the mutable expiration time, in contrast to the immutable expiration time represented by the `CancelAfter` field. The `Expiration` field is always omitted when a `PayChannel` node is created. There are several ways the `Expiration` field of a `PayChannel` node can be updated. Generally, the source address can set the `Expiration` of a channel freely  ...
```
> There are several ways the `Expiration` field of a `PayChannel` node can be updated.

What are they? I think you explain them in the rest of the section, but it might be nice to have a link between this explanation and a clear list of options. Maybe add a couple of headings to make it clear who can do what under what conditions? I know a "Destination Address" section would be very short, but might help readers grok the overview a little easier.

mDuo13 replied:
The rest of this section is the clearest list of options I could come up with. I rearranged this paragraph and added some h4s to the rest of the section to make that clearer.

ryangyoung commented on content/reference-ledger-format.md:
```diff
@@ -303,6 +305,97 @@ The `index` of an Offer node is the SHA-512Half of the following values put toge
+
+The `Expiration` field of a payment channel is the mutable expiration time, in contrast to the immutable expiration time represented by the `CancelAfter` field. The `Expiration` field is always omitted when a `PayChannel` node is created. There are several ways the `Expiration` field of a `PayChannel` node can be updated. Generally, the source address can set the `Expiration` of a channel freely  ...
+
+The source address can set the `Expiration` directly with the PaymentChannelFund transaction type. The new value must not be earlier than whichever of the following values is earliest:
```
> the PaymentChannelFund transaction type

link?

mDuo13 replied:
This refers to content from the other PR, so I can't add the link until both PRs are merged.

ryangyoung reviewed (approved):
Very clear explanation of a very complicated feature! Left a couple of suggestions for some changes that might make the complexities a bit more digestible. Those would be nice, but not totally required for the merge (especially the diagram).

mDuo13 committed:
paychan ledger node - edits per Ryan's re-review
